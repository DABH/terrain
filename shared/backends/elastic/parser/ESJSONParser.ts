/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import ESJSONType from './ESJSONType';
import ESParameter from './ESParameter';
import ESParserError from './ESParserError';
import ESParserToken from './ESParserToken';
import ESPropertyInfo from './ESPropertyInfo';
import ESValueInfo from './ESValueInfo';

/**
 * An instrumented JSON parser for parsing ES queries
 *
 * http://www.json.org/
 *
 * https://discuss.codemirror.net/t/custom-mode-any-way-to-process-the-entire-string-in-the-editor-before-running-token-stream-state/775
 * "It is usually less work (and less of a performance hazard) to implement a
 * regular streaming CodeMirror mode than to try and work on top of a classical
 * whole-file parser. But if you really want to do it this way, you can
 * parameterize your mode with a full token stream, including token text,
 * and have it run through that (keeping a counter in its state) as long as
 * the stream it has matches the content, and just start outputting null
 * tokens when it runs into an outdated token. You can then set a debounced
 * change event handler that parses the document and calls cm.setOption("mode",
 * myModeWithTokens(tokenizeWholeFile(cm.getValue())), starting a new highlight."
 *
 */
export default class ESJSONParser
{
  private queryString: string; // string being parsed
  private allowParameters: boolean; // if @parameters are allowed (non-compliant)

  private charNumber: number; // current parser/tokenizer position in the queryString

  private lastCheckedRowChar: number; // last char checked for row count
  private lastRowChar: number; // position of last newline
  private lastRowNumber: number; // row number of last newline

  private value: any; // parsed value
  private valueInfo: ESValueInfo | null; // parsed root value info (or null on failure)

  private tokens: ESParserToken[]; // accumulated tokens, in order
  private valueInfos: ESValueInfo[]; // accumulated value info's, in order
  private errors: ESParserError[]; // accumulated errors, in order

  /**
   * Runs the parser on the given query string. Read needed data by calling the
   * public member functions below.
   * @param queryString query to parse
   * @param allowParameters
   */
  public constructor(queryString: string, allowParameters: boolean = true)
  {
    this.queryString = queryString;
    this.charNumber = 0;
    this.lastCheckedRowChar = -1;
    this.lastRowChar = 0;
    this.lastRowNumber = 0;
    this.value = null;
    this.valueInfo = null;
    this.allowParameters = allowParameters;

    this.tokens = [];
    this.valueInfos = [];
    this.errors = [];

    this.valueInfo = this.readValue();
    if (this.valueInfo !== null)
    {
      this.value = this.valueInfo.value;
    }

    // check ending
    if (this.peek() !== '')
    {
      this.accumulateToken();
      this.accumulateErrorOnCurrentToken('Unexpected token at the end of the query string');
    }
  }

  /**
   * @returns {any} the parsed root value
   */
  public getValue(): any
  {
    return this.value;
  }

  /**
   * @returns {any} the parsed root value info
   */
  public getValueInfo(): ESValueInfo
  {
    return this.valueInfo as ESValueInfo;
  }

  /**
   * @returns {ESParserToken[]} the tokens encountered, in order
   */
  public getTokens(): ESParserToken[]
  {
    return this.tokens;
  }

  /**
   * @returns {ESValueInfo[]} metadata value info's for each value parsed, in order
   */
  public getValueInfos(): ESValueInfo[]
  {
    return this.valueInfos;
  }

  /**
   * @returns {ESParserError[]} errors encountered, in order
   */
  public getErrors(): ESParserError[]
  {
    return this.errors;
  }

  public hasError(): boolean
  {
    return this.errors.length > 0;
  }

  public accumulateError(error: ESParserError): void
  {
    this.errors.push(error);
  }

  private peek(): string
  {
    // skip whitespace
    this.match(/^\s*/);

    // handle EOF
    if (this.charNumber >= this.queryString.length)
    {
      return '';
    }

    return this.queryString.charAt(this.charNumber);
  }

  private peekString(): string
  {
    if (this.charNumber >= this.queryString.length)
    {
      return '';
    }

    return this.queryString.substr(this.charNumber, this.queryString.length);
  }

  private advance(numChars: number = 1): void
  {
    this.charNumber += numChars;
  }

  private readValue(): ESValueInfo | null
  {
    const valueInfo: ESValueInfo = this.beginValueInfo();
    const token: ESParserToken = this.accumulateToken();

    const nextChar: string = this.peek();

    let jsonType: ESJSONType = ESJSONType.invalid;

    try
    {
      switch (nextChar)
      {
        // string
        case '"':
          jsonType = ESJSONType.string;
          valueInfo.value = this.readString();
          break;

        // number
        case '0':
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
        case '-':
          jsonType = ESJSONType.number;
          valueInfo.value = this.readNumber();
          break;

        // object
        case '{':
          jsonType = ESJSONType.object;
          this.advance();
          this.readObject(valueInfo);
          break;

        // object end
        case '}':
          break;

        // array
        case '[':
          jsonType = ESJSONType.array;
          this.advance();
          this.readArray(valueInfo);
          break;

        // array end
        case ']':
          break;

        // true
        case 't':
          jsonType = ESJSONType.boolean;
          valueInfo.value = this.readTrueValue();
          break;

        // false
        case 'f':
          jsonType = ESJSONType.boolean;
          valueInfo.value = this.readFalseValue();
          break;

        // null
        case 'n':
          jsonType = ESJSONType.null;
          valueInfo.value = this.readNullValue();
          break;

        case '@':
          jsonType = ESJSONType.parameter;
          valueInfo.value = this.readParameter();
          break;

        default:
          this.accumulateErrorOnCurrentToken('Unknown token found when expecting a value');
          this.matchAndSetToken(/^.[a-zA-Z_0-9]*/); // try to skip the token
          break;
      }
    }
    finally
    {
      if (valueInfo.value === undefined)
      {
        // if no value was read, erase the token information accumulated
        this.valueInfos.pop();
        this.tokens.pop();
      }
    }

    if (valueInfo.value === undefined)
    {
      return null;
    }

    token.jsonType = jsonType;
    valueInfo.jsonType = jsonType;

    return valueInfo;
  }

  private readString(): string
  {
    let result: any = this.captureMatch(/^("(?:\\(?:["\\\/bfnrt]|u[a-fA-F0-9]{4})|[^"\\\0-\x1F\x7F])*")/);
    if (typeof result === 'string')
    {
      return result;
    }

    // try to capture an invalid token
    this.accumulateErrorOnCurrentToken('Unknown string format');

    // try to capture a string that ends in double quotes
    result = this.matchAndSetToken(/^"(?:\\.|[^"\\])*"/);
    if (result !== null)
    {
      return result.substring(1, result.length - 1);
    }

    // try to capture a string that ends with some JSON control char
    result = this.matchAndSetToken(/^".+[^,:\[\]{}"]/);

    if (result !== null)
    {
      return result.substring(1, result.length);
    }

    this.advance();
    return '';
  }

  private readNumber(): number
  {
    const result: any = this.captureMatch(/^(-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?)/);
    if (typeof result === 'number')
    {
      return result;
    }

    // try to capture an invalid token
    this.accumulateErrorOnCurrentToken('Unknown number format');
    this.matchAndSetToken(/^([\-.eE0-9]+)/);
    return 0;
  }

  private readArray(arrayInfo: ESValueInfo): void
  {
    const array: any[] = [];
    arrayInfo.value = array;
    arrayInfo.arrayChildren = [];

    for (let elementInfo: ESValueInfo | null = this.readValue();
      elementInfo !== null;
      elementInfo = this.readValue())
    {
      array.push(elementInfo.value);
      arrayInfo.arrayChildren.push(elementInfo);

      // read array delimiter, ','
      const propertyDelimiter: string = this.peek();
      if (propertyDelimiter !== ',')
      {
        break;
      }

      this.accumulateToken(arrayInfo, ESJSONType.arrayDelimiter);
      this.advance();
    }

    // at the end of the list, make sure that the terminating char is ']'
    if (this.peek() === ']')
    {
      this.accumulateToken(arrayInfo, ESJSONType.arrayTerminator);
      this.advance();
    }
    else
    {
      this.accumulateErrorOnCurrentToken('Missing or misplaced array closing bracket, "]"');
    }
  }

  private readObject(objInfo: ESValueInfo): void
  {
    const obj: object = {};
    objInfo.value = obj;
    objInfo.objectChildren = {};

    for (let nameInfo: ESValueInfo | null = this.readValue();
      nameInfo !== null;
      nameInfo = this.readValue())
    {
      let propertyName: any = nameInfo.value;

      if (typeof propertyName !== 'string')
      {
        this.accumulateErrorOnCurrentToken(
          'Object property names must be strings, but found a ' + typeof propertyName + ' instead');
        propertyName = String(propertyName);
      }

      // check for duplicates
      if (obj.hasOwnProperty(propertyName))
      {
        this.accumulateErrorOnCurrentToken('Duplicate property names are not allowed');
      }

      // install a property info into the parent object
      const propertyInfo: ESPropertyInfo = new ESPropertyInfo(nameInfo);
      objInfo.objectChildren[propertyName] = propertyInfo;

      // read delimiter between property name and value
      const kvpDelimiter: string = this.peek();

      // check for errors in the delimiter
      if (kvpDelimiter !== ':')
      {
        if (kvpDelimiter === ',')
        {
          // eat object delimiter, ','
          this.accumulateErrorOnCurrentToken('Object property\'s value is missing');
          this.accumulateToken(objInfo, ESJSONType.objectDelimiter);
          this.advance();
          continue;
        }

        break;
      }

      // eat property delimiter, ':'
      this.accumulateToken(nameInfo, ESJSONType.propertyDelimiter);
      this.advance();

      // read property value
      const propertyValueInfo: ESValueInfo | null = this.readValue();

      // check for errors in the property value
      if (propertyValueInfo === null)
      {
        this.accumulateErrorOnCurrentToken('Object property\'s value is missing');
        break;
      }

      propertyInfo.propertyValue = propertyValueInfo;
      obj[propertyName] = propertyValueInfo.value; // set object property

      // read next delimiter
      const propertyDelimiter: string = this.peek();
      if (propertyDelimiter !== ',')
      {
        break;
      }

      // eat object delimiter, ','
      this.accumulateToken(objInfo, ESJSONType.objectDelimiter);
      this.advance();
    }

    // at the end of the object, make sure that the terminating char is '}'
    if (this.peek() === '}')
    {
      // eat object terminator, ','
      this.accumulateToken(objInfo, ESJSONType.objectTerminator);
      this.advance();
    }
    else
    {
      this.accumulateErrorOnCurrentToken('Missing or misplaced object closing brace, "}"');
    }
  }

  private readTrueValue(): boolean
  {
    this.readBooleanOrNull(/^true(?!\w)/);
    return true;
  }

  private readFalseValue(): boolean
  {
    this.readBooleanOrNull(/^false(?!\w)/);
    return false;
  }

  private readNullValue(): null
  {
    this.readBooleanOrNull(/^null(?!\w)/);
    return null;
  }

  private readBooleanOrNull(exp: RegExp): boolean
  {
    const match: string | null = this.matchAndSetToken(exp);
    if (match !== null)
    {
      return true;
    }

    // try to capture an invalid token
    this.accumulateErrorOnCurrentToken('Unknown value type, possibly a boolean null (true, false, and null, are valid).');
    this.matchAndSetToken(/^\w+/);
    return false;
  }

  private readParameter(): ESParameter
  {
    let match: string | null = this.matchAndSetToken(/^@([a-zA-Z_][a-zA-Z_0-9]*)/);
    if (match === null)
    {
      match = '';
      this.accumulateErrorOnCurrentToken(
        'Invalid parameter name. Parameter names must begin with a letter or underscore, ' +
        'and can only contain letters, underscores, and numbers.');
    }

    return new ESParameter(match);
  }

  private captureMatch(exp: RegExp): any
  {
    const match: string | null = this.matchAndSetToken(exp);
    if (match === null)
    {
      return null;
    }

    return JSON.parse(match);
  }

  private matchAndSetToken(exp: RegExp): string | null
  {
    const result: string | null = this.match(exp);
    if (result !== null)
    {
      this.setToken();
    }

    return result;
  }

  private match(exp: RegExp): string | null
  {
    const matches = this.peekString().match(exp);
    if (matches === null || matches.length <= 0)
    {
      return null;
    }

    const match: string = matches[0];
    this.advance(match.length);
    return match;
  }

  private accumulateToken(valueInfo: ESValueInfo | null = null, jsonType: ESJSONType | null = null): ESParserToken
  {
    const element: ESParserToken =
      new ESParserToken(this.charNumber,
        this.getRow(),
        this.getCol(),
        1,
        this.queryString.substring(this.charNumber, this.charNumber + 1));

    if (jsonType !== null)
    {
      element.jsonType = jsonType;
    }

    this.tokens.push(element);

    // link up the ValueInfo and the Token
    if (valueInfo === null)
    {
      valueInfo = this.getCurrentValueInfo();
    }

    // element.valueInfo = valueInfo;

    if (valueInfo !== null)
    {
      valueInfo.tokens.push(element);
    }

    return element;
  }

  private getCurrentToken(): ESParserToken
  {
    return this.tokens[this.tokens.length - 1];
  }

  private setToken(): void
  {
    const element: ESParserToken = this.getCurrentToken();
    element.length = this.charNumber - element.charNumber;
    element.toRow = this.getRow();
    element.toCol = this.getCol();
    element.substring = this.queryString.substring(element.charNumber, element.charNumber + element.length);
  }

  private beginValueInfo(): ESValueInfo
  {
    const element: ESValueInfo = new ESValueInfo();
    this.valueInfos.push(element);
    return element;
  }

  private getCurrentValueInfo(): ESValueInfo | null
  {
    if (this.valueInfos.length === 0)
    {
      return null;
    }

    return this.valueInfos[this.valueInfos.length - 1];
  }

  private accumulateErrorOnCurrentToken(message: string): void
  {
    this.errors.push(new ESParserError(
      this.getCurrentToken(), this.getCurrentValueInfo() as ESValueInfo, message));
  }

  private getRow(): number
  {
    while (this.lastCheckedRowChar < this.charNumber)
    {
      ++this.lastCheckedRowChar;
      const c: string = this.queryString.charAt(this.lastCheckedRowChar);
      if (c === '\n')
      {
        this.lastRowChar = this.lastCheckedRowChar + 1;
        ++this.lastRowNumber;
      }
    }

    return this.lastRowNumber;
  }

  private getCol(): number
  {
    this.getRow();
    return this.charNumber - this.lastRowChar;
  }
}
