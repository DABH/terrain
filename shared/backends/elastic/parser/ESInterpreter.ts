/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import EQLConfig from './EQLConfig';
import ESJSONParser from './ESJSONParser';
import ESParserError from './ESParserError';
import ESValueInfo from './ESValueInfo';

export const ESInterpreterDefaultConfig = new EQLConfig();

/**
 * An instrumented interpreter that takes the output of ESJSONParser and
 * decorates the results with ES-specific information.
 */
export default class ESInterpreter
{
  public parser: ESJSONParser; // source parser
  public config: EQLConfig; // query language description

  /**
   * Runs the interpreter on the given query string. Read needed data by calling the
   * public member functions below. You can also pass in an existing ESJSONParser
   * to run the interpreter on it's result.
   * @param query the query string or parser to interpret
   */
  public constructor(query: string | ESJSONParser, config: EQLConfig = ESInterpreterDefaultConfig)
  {
    this.config = config;

    if (typeof query === 'string')
    {
      this.parser = new ESJSONParser(query);
    } else
    {
      this.parser = query;
    }

    if (this.parser.getValue() !== null && this.parser.hasError() === false)
    {
      try
      {
        this.config.getClause('root').mark(this, this.parser.getValueInfo());
      } catch (e)
      {
        this.accumulateError(this.parser.getValueInfo(), 'Failed to mark the json object ' + String(e.message));
      }
    }
  }

  public accumulateError(info: ESValueInfo, message: string, isWarning: boolean = false): void
  {
    this.parser.accumulateError(new ESParserError(info.tokens[0], info, message, isWarning));
  }

  public hasError()
  {
    return this.parser === null || this.parser.hasError();
  }
}
