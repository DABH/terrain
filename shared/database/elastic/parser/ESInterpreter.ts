/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import ESParser from 'shared/database/elastic/parser/ESParser';
import ESClause from './clauses/ESClause';
import EQLConfig from './EQLConfig';
import ESJSONParser from './ESJSONParser';
import ESParserError from './ESParserError';
import ESValueInfo from './ESValueInfo';

export const ESInterpreterDefaultConfig = new EQLConfig();

/**
 * An instrumented interpreter that takes the output of ESJSONParser and
 * decorates the results with ES-specific information.
 */
export default class ESInterpreter
{
  public config: EQLConfig; // query language description
  public params: { [name: string]: null | ESClause }; // input parameter clause types
  public parser: ESParser; // source parser
  public rootValueInfo: ESValueInfo;
  public errors: ESParserError[];

  /**
   * Runs the interpreter on the given query string. Read needed data by calling the
   * public member functions below. You can also pass in an existing ESJSONParser
   * to run the interpreter on it's result.
   *
   * 1) parse
   * 2) interpret
   *
   * @param query the query string or parser to interpret
   * @param config the spec config to use
   * @param params parameter map to use
   */
  public constructor(query: string | ESParser,
    params: { [name: string]: any } = {},
    config: EQLConfig = ESInterpreterDefaultConfig)
  {
    this.config = config;
    this.params = params;
    this.errors = [];

    if (typeof query === 'string')
    {
      this.parser = new ESJSONParser(query) as ESParser;
      if (this.parser.hasError())
      {
        this.accumulateError(null, 'Failed to parse the query ' + query);
        return;
      }
      this.rootValueInfo = this.parser.getValueInfo();
    } else if (query instanceof ESParser)
    {
      this.parser = query;
      this.rootValueInfo = this.parser.getValueInfo();
    } else
    {
      this.parser = null;
      this.accumulateError(null, 'The input must be a query, an ESJSONParser object, or an ESValueInfo');
      return;
    }

    try
    {
      const root: ESValueInfo = this.rootValueInfo;
      if (root.clause === undefined)
      {
        root.clause = this.config.getClause('root');
      }
      root.recursivelyVisit(
        (info: ESValueInfo): boolean =>
        {
          if (info.parameter !== undefined)
          {
            const value: null | any = this.params[info.parameter];
            if (value === undefined)
            {
              this.accumulateError(info, 'Undefined parameter: ' + info.parameter);
            }

            return false; // don't validate parameters
          }

          if (info.clause !== undefined)
          {
            info.clause.mark(this, info);
          }

          return true;
        });
    } catch (e)
    {
      this.accumulateError(this.rootValueInfo, 'Failed to mark the json object ' + String(e.message));
    }
  }

  public accumulateError(info: ESValueInfo, message: string, isWarning: boolean = false): void
  {
    let token = null;
    if (info !== null && info !== undefined && info.tokens.length > 0)
    {
      token = info.tokens[0];
    }
    const e = new ESParserError(token, info, message, isWarning);
    this.errors.push(e);
    if (this.parser !== null)
    {
      this.parser.accumulateError(e);
    }
  }

  public hasInterpretingError()
  {
    return this.errors.length > 0;
  }

  public hasJSONParsingError()
  {
    return this.parser === null || this.parser.hasError();
  }

  public hasError()
  {
    return this.hasInterpretingError() || this.hasJSONParsingError();
  }
}
