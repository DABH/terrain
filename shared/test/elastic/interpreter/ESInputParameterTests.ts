/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.

import * as fs from 'fs';
import ESInterpreter from 'shared/database/elastic/parser/ESInterpreter';
import * as winston from 'winston';
import ESJSONParser from '../../../database/elastic/parser/ESJSONParser';
import { makePromiseCallback } from '../../Utils';

import * as Immutable from 'immutable';
import ESParserError from 'shared/database/elastic/parser/ESParserError';

function getExpectedFile(): string
{
  return __filename.split('.')[0] + '.expected';
}

let expected;

beforeAll(async (done) =>
{
  // TODO: get rid of this monstrosity once @types/winston is updated.
  (winston as any).level = 'debug';

  const contents: any = await new Promise((resolve, reject) =>
  {
    fs.readFile(getExpectedFile(), makePromiseCallback(resolve, reject));
  });
  expected = JSON.parse(contents);
  done();
});

function testCardParseWithInputParameter(testName: string,
  testString: string,
  expectedValue: any,
  inputParameter: any,
  expectedErrors: ESParserError[] = [])
{
  winston.info('testing "' + testName + '": "' + testString + '"' + '": "' + JSON.stringify(inputParameter) + '"');
  const emptyCards = Immutable.List([]);
  const parameters = { number: 10 };
  const interpreter: ESInterpreter = new ESInterpreter(testString, parameters);
  const parser: ESJSONParser = interpreter.parser as ESJSONParser;
  const rootValueInfo = parser.getValueInfo();
  expect(rootValueInfo.value).toEqual(expectedValue);
  expect(interpreter.errors).toEqual(expectedErrors);
}

test('parse card', () =>
{
  Object.getOwnPropertyNames(expected).forEach(
    (testName: string) =>
    {
      const testValue: any = expected[testName];
      const testString = testValue['query'];
      const inputParameter = testValue['parameter'];
      const expectedValue = testValue['expect'];
      testCardParseWithInputParameter('test', testString, expectedValue, inputParameter);
    });
});
