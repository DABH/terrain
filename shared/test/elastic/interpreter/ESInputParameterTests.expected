{
  "empty query" : {
    "query": "{}",
    "expect": {},
    "parameter": "{}"
  },
  "number input" : {
    "query": "{\"from\": @number}",
    "expect": {
      "from" : "@number"
    },
    "parameter": {"number": 0}
  },
  "runtime input" : {
    "query": "{\"groupJoin\": {\"test\": {\"from\": @parent}}}",
    "expect": {
      "groupJoin": {
        "test": {
          "from" : "@parent"
        }
      }
    },
    "parameter": {}
  },
  "runtime input with parent alias" : {
    "query": "{\"groupJoin\": {\"parentAlias\": \"user\", \"test\": {\"from\": @user}}}",
    "expect": {
      "groupJoin": {
        "parentAlias": "user",
        "test": {
          "from" : "@user"
        }
      }
    },
    "parameter": {}
  }
}
